QueryEngine/ArithmeticIR.cpp

```diff
@@ -660,3 +572,69 @@ llvm::Value* CodeGenerator::codegenUMinus(const Analyzer::UOper* uoper,
   }
   return ret;
 }
+
+llvm::Value* CodeGenerator::codegenBinOpWithOverflowCheck(
+    const Analyzer::BinOper* bin_oper,
+    llvm::Value* lhs_lv,
+    llvm::Value* rhs_lv,
+    const std::string& null_check_suffix,
+    const SQLTypeInfo& ti) {
+  cgen_state_->needs_error_check_ = true;
+
+  std::string bb_prefix;
+  llvm::Intrinsic::ID fn_id;
+  switch (bin_oper->get_optype()) {
+    case kMINUS:
+      bb_prefix = "sub";
+      fn_id = llvm::Intrinsic::ssub_with_overflow;
+      break;
+    case kPLUS:
+      bb_prefix = "add";
+      fn_id = llvm::Intrinsic::sadd_with_overflow;
+      break;
+    case kMULTIPLY:
+      bb_prefix = "mul";
+      fn_id = llvm::Intrinsic::smul_with_overflow;
+      break;
+    default:
+      CHECK(false);
```

Comment:
We're gradually replacing instances of `CHECK(false)` with `LOG(FATAL)` with a meaningful error message. Also, you might need to initialize `fn_id` if we can't make the compiler see that the path through `default` is unreachable, to squash the warning. Maybe move this `switch` to a helper function as well, it's easy to isolate and makes this method smaller.

Reply:
Tanks for comments! I'll rework this part.